runtime/tutorials/chat_app.md

```diff
@@ -163,73 +140,154 @@ function broadcast_usernames() {
-  const socket = await ctx.upgrade();
-  const username = ctx.request.url.searchParams.get("username");
-  if (connectedClients.has(username)) {
-    socket.close(1008, `Username ${username} is already taken`);
-    return;
+We'll set up the client side JavaScript in an `app.js` file, you'll have seen it
+linked in the HTML we just wrote. In the `public` folder and add an `app.js`
+file with the following code:
+
+```js
+// Initialize WebSocket connection
+const myUsername = prompt("Please enter your name") || "Anonymous";
+const socket = new WebSocket(
+  `ws://localhost:8080/start_web_socket?username=${myUsername}`,
+);
+
+// Handle WebSocket messages
+socket.onmessage = (event) => {
+  const data = JSON.parse(event.data);
+
+  switch (data.event) {
+    case "update-users":
+      updateUserList(data.usernames);
+      break;
+
+    case "send-message":
+      addMessage(data.username, data.message);
+      break;
   }
-  socket.username = username;
-  connectedClients.set(username, socket);
-  console.log(`New client connected: ${username}`);
-
-  // broadcast the active users list when a new user logs in
-  socket.onopen = () => {
-    broadcast_usernames();
-  };
-
-  // when a client disconnects, remove them from the connected clients list
-  // and broadcast the active users list
-  socket.onclose = () => {
-    console.log(`Client ${socket.username} disconnected`);
-    connectedClients.delete(socket.username);
-    broadcast_usernames();
-  };
-
-  // broadcast new message if someone sent one
-  socket.onmessage = (m) => {
-    const data = JSON.parse(m.data);
-    switch (data.event) {
-      case "send-message":
-        broadcast(
-          JSON.stringify({
-            event: "send-message",
-            username: socket.username,
-            message: data.message,
-          }),
-        );
-        break;
-    }
-  };
-});
+};
 
-app.use(router.routes());
-app.use(router.allowedMethods());
-app.use(async (context) => {
-  await context.send({
-    root: `${Deno.cwd()}/`,
-    index: "public/index.html",
+// Update user list in the DOM
+function updateUserList(usernames) {
+  const userList = document.getElementById("users");
+  userList.innerHTML = ""; // Clear existing list
+
+  for (const username of usernames) {
+    const listItem = document.createElement("li");
+    listItem.textContent = username;
+    userList.appendChild(listItem);
+  }
+}
+
+// Add a new message to the conversation
+function addMessage(username, message) {
+  const conversation = document.getElementById("conversation");
+  const messageDiv = document.createElement("div");
+  messageDiv.innerHTML = `<span>${username}</span> ${message}`;
+  conversation.prepend(messageDiv);
+}
+
+// Set up event listener for sending messages
+window.onload = () => {
+  const inputElement = document.getElementById("data");
+  const sendButton = document.getElementById("send");
+
+  // Listen for Enter key press
+  inputElement.addEventListener("keypress", (e) => {
```

Comment:
The `keypress` event has been [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Element/keypress_event) for a while.

Reply:
Updated